inc/system/diagnostic/databasekeyschecker.class.php

```diff
@@ -0,0 +1,247 @@
+<?php
+/**
+ * ---------------------------------------------------------------------
+ * GLPI - Gestionnaire Libre de Parc Informatique
+ * Copyright (C) 2015-2021 Teclib' and contributors.
+ *
+ * http://glpi-project.org
+ *
+ * based on GLPI - Gestionnaire Libre de Parc Informatique
+ * Copyright (C) 2003-2014 by the INDEPNET Development Team.
+ *
+ * ---------------------------------------------------------------------
+ *
+ * LICENSE
+ *
+ * This file is part of GLPI.
+ *
+ * GLPI is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * GLPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+namespace Glpi\System\Diagnostic;
+
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access this file directly");
+}
+
+use DBmysql;
+
+/**
+ * @since x.x.x
+ */
+class DatabaseKeysChecker {
+
+   /**
+    * DB instance.
+    *
+    * @var DBmysql
+    */
+   protected $db;
+
+   /**
+    * Local cache for tables columns.
+    *
+    * @var array
+    */
+   private $columns = [];
+
+   /**
+    * Local cache for tables indexes.
+    *
+    * @var array
+    */
+   private $indexes = [];
+
+   /**
+    * @param DBmysql $db   DB instance.
+    */
+   public function __construct(DBmysql $db) {
+      $this->db = $db;
+   }
+
+   /**
+    * Get list of missing keys, basing detection on column names.
+    *
+    * Array keys are expected key names, and array values are fields that should be contained in these keys.
+    *
+    * @param string $table_name
+    *
+    * @return array
+    */
+   public function getMissingKeys(string $table_name): array {
+      $missing_keys = [];
+
+      $misnamed_keys = $this->getMisnamedKeys($table_name);
+
+      $columns = $this->getColumnsNames($table_name);
+      foreach ($columns as $column_name) {
+         $column_name_matches = [];
+         if (preg_match('/^items_id(?<suffix>_.+)?/', $column_name, $column_name_matches)) {
+            $suffix = $column_name_matches['suffix'] ?? '';
+            $expected_key = 'item' . $suffix;
+            if (in_array('itemtype' . $suffix, $columns)
+               && !in_array($expected_key, $misnamed_keys)
+               && !$this->areFieldsCorrecltyIndexed($table_name, ['itemtype' . $suffix, 'items_id' . $suffix], $expected_key)
+               && !$this->areFieldsCorrecltyIndexed($table_name, ['items_id' . $suffix, 'itemtype' . $suffix], $expected_key)) {
```

Comment:
Should we accept reverse order (i.e. ```(`items_id`,`itemtype`)```) ?
I guess it could happen to have a query that only filters on `itemtype`, but it should not happen to have a query that only filters on `items_id`. Only `glpi_registeredids` table have this reverse order.

Reply:
glpi_registeredids seems to be a specific case; reverse order is not mandatory

Reply:
I changed order of this index.
I also removed other keys in the same order, as we had some tables that had keys on both `itemtype,items_id` and `items_id,itemtype`.